walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,115 @@
 
 package com.google.sps;
 
+import com.google.common.collect.ImmutableList; 
+import com.google.common.collect.ImmutableSet; 
+import com.google.common.collect.Sets; 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set; 
+import java.util.stream.Collectors; 
 
+/** A class to determine appropriate meeting times */
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+  private static final int LENGTH_OF_DAY = 24 * 60; 
+
+  /** 
+  * Returns a list of time ranges such that a meeting with the specified constraints can be held
+  * within each time range while still respecting the event calendar 
+  */
+  public static Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+    
+    long meetingDuration = request.getDuration(); 
+    Collection<String> mandatoryMeetingAttendees = request.getAttendees();
+    Collection<String> optionalMeetingAttendees = request.getOptionalAttendees();
+
+    // Return no availability if the requested meeting's duration is longer than one day
+    if (meetingDuration > (LENGTH_OF_DAY)) {
+        return Arrays.asList();
+    }
+
+    ImmutableList<TimeRange> mandatoryRanges = determineAvailability(events, mandatoryMeetingAttendees, meetingDuration);
+    ImmutableList<TimeRange> optionalRanges = determineAvailability(events, optionalMeetingAttendees, meetingDuration);
+    
+    // Determine if availability is dependent upon solely mandatory verus optional attendees 
+    if (mandatoryMeetingAttendees.isEmpty() && optionalMeetingAttendees.isEmpty()) {
+        // Return availability as the entire day if there are no mandatory nor optional attendees 
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    } if (!mandatoryMeetingAttendees.isEmpty() && optionalMeetingAttendees.isEmpty()) {
+        return mandatoryRanges; 
+    } if (mandatoryMeetingAttendees.isEmpty() && !optionalMeetingAttendees.isEmpty()) {
+        return optionalRanges; 
+    } 
+        
+    // Merge mandatory and optional attendees 
+    List<String> mergedAttendees = new ArrayList<>(mandatoryMeetingAttendees);
```

Comment:
Is order of attendees important here? If so, use ImmutableList, if not ImmutableSet, 

```java
ImmutableList.<String>builder() // note placement of period between List, <String>
  .addAll(mandatoryMeetingAttendees)
  .addAll(optionalMeetingAttendees)
  .build();
```

Reply:
changed to ImmutableSet